### PR TITLE
feat(module:upload): new events & bug fix

### PR DESCRIPTION
**Pull request by anddrzejb**
### 🤔 This is a ...

- [x] New feature
- [x] Bug fix
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Component style update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Refactoring
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Other (about what?)

### 🔗 Related issue link
The code below, when adding a new file, will throw `NullReferenceException`.
```razor
@inject MessageService _message

<Upload @attributes="attrs"
        OnSingleCompleted="OnSingleCompleted"
        OnCompleted="HandleFileListChange">
    <Button Icon="upload">
        <span>点击上传</span>
    </Button>
</Upload>
@code
{
    Upload upload;

    Dictionary<string, object> attrs = new Dictionary<string, object>
{
        {"Action", "https://www.mocky.io/v2/5cc8019d300000980a055e76" },
        {"Name", "files" },
        {"DefaultFileList", new List<UploadFileItem>
            {
                new UploadFileItem
                {
                    Id = "1",
                    FileName = "1.jpg",
                    Url = "https://www.baidu.com/1.jpg",
                    State = UploadState.Success
                }
            }
        }
    };

    void OnSingleCompleted(UploadInfo fileinfo)
    {
        if (fileinfo.File.State == UploadState.Success)
        {
            var result = fileinfo.File.GetResponse<ResponseModel>();
            fileinfo.File.Url = result.url;
        }
    }

    void HandleFileListChange(UploadInfo fileinfo)
    {
        Console.WriteLine($"HandleFileListChange: {fileinfo.File.FileName}, FileList count: {fileinfo.FileList.Count}");
    }

    public class ResponseModel
    {
        public string name { get; set; }

        public string status { get; set; }

        public string url { get; set; }

        public string thumbUrl { get; set; }
    }
}
```

### 💡 Background and solution
1. Added missing event `OnDownload` (exists in react antDesign)
2. Added 2 new events `BeforeAllUpload` and its async counterpart. Use case - to check whole upload list for duplicates, instead go file-by-file using `BeforeUpload`.
3. Bug fix (mentioned in _Related Issue_ : the new file list was not assigned to local variable.
4. Typescript functions got type checked arguments for file upload. 

### ☑️ Self Check before Merge

⚠️ Please check all items below before review. ⚠️

- [x] Doc is updated/provided or not needed
- [x] Demo is updated/provided or not needed
- [x] Changelog is provided or not needed

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=1297) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=1297) before we can accept your contribution.<br/><br/>:x: anddrzejb [sign now](https://cla.dotnetfoundation.org/ant-design-blazor/ant-design-blazor?pullRequest=1297)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/ant-design-blazor/ant-design-blazor?pullRequest=1297) it.</sub>

File: components/table/ActionColumn.razor [hunk 1]
```diff
@@ -26,13 +31,42 @@ else if (IsColGroup)
...
+        <th class="ant-table-cell ant-table-row-expand-icon-cell"></th>
+    }
+
     <th class="@ClassMapper.Class" style="@FixedStyle @HeaderStyle" colspan="@HeaderColSpan">
         @if (TitleTemplate != null)@TitleTemplate else @Title
     </th>
 }
 else if (!IsHeader && RowSpan != 0 && ColSpan != 0)
 {
+    @if (AppendExpandColumn)
```
Comment (ElderJames) [inline components/table/ActionColumn.razor:45]: You're bringing other commits to this PR.

**Review by ElderJames — CHANGES_REQUESTED** [review]

**Comment by ElderJames** [comment]
Welcome back! 😊

File: components/table/ActionColumn.razor (hunk 1 above)
Comment (anddrzejb) [reply components/table/ActionColumn.razor:45]: Crap... I was doing this on master, than I realized this is should be feature... I will redo this PR from feature branch.

**Pull request closed without merging**